### PR TITLE
Feature/initial solution

### LINKED ISSUE #1: Create solution PoC
Design and create PoC.

### PR DESCRIPTION
**Pull request by miroslavpojer**
- Initial implementation of the solution.

Now, we can extend our IT test approach with this GH action, which can check our repositories for forgotten test files not following the test file naming convention.

Closes #1

### CODE REVIEW

**Comment by miroslavpojer** [comment]
After discussing with @lsulak, I have migrated this solution to a 100% GH Action composite solution to remove dependency on Javascript and keep the solution 100% Python-based.

**Comment by miroslavpojer** [comment]
Release Notes
- Designed and implemented GH Composite Action.

**Comment by miroslavpojer** [comment]
> After discussing with @lsulak, I have migrated this solution to a 100% GH Action composite solution to remove dependency on Javascript and keep the solution 100% Python-based.

See experiment workflow file - ready to run:
- https://github.com/absa-group/living-doc-example-project/actions/runs/9265458006/workflow

File: README.md [hunk 1]
```diff
@@ -1,2 +1,172 @@
...
+  - [Default](#default)
+  - [Full example](#full-example)
+- [Running Unit Tests](#running-unit-tests)
+- [Code Coverage](#code-coverage)
+- [Run Action Locally](#run-action-locally)
+- [Contributing](#contributing)
+- [License](#license)
+
+## Description
+The **Test File Suffix Inspector** GitHub Action is designed to ensure naming conventions in test files within specified repository directories. It scans for test files and reports any missing specified suffixes, helping maintain consistency and adherence to project standards.
```
Comment (Zejnilovic) [inline README.md:18]: It is a scala specific tool, o so it seems, but I don't see that here

File: README.md [hunk 2]
```diff
@@ -1,2 +1,172 @@
...
+      runs-on: ubuntu-latest
+      steps:
+        - uses: actions/checkout@v2
+
+        - name: Test File Suffix Inspector Full Custom
+          id: scan-test-files
+          uses: AbsaOSS/test-file-suffix-inspector@v0.1.0
+          with:
+            suffixes: 'UnitTests,IntegrationTests'
+            include_directories: 'src/test/'
```
Comment (Zejnilovic) [inline README.md:117]: How does this work? Does it do `src/test` per module so it is always like `{module}/src/test/*`. 
Does it use glob patterns?

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+coverage
```
Comment (Zejnilovic) [inline requirements.txt:1]: no specific version?

File: README.md (hunk 2 above)
Comment (miroslavpojer) [reply README.md:117]: The action uses the project's current working dir as a starting point. Then, it scans for all folders on the current level, removes dir exclusions, excludes by file exclusions, and checks for file patterns.

The logic does not expect any project structure. It is oriented only on folders and files.

File: .github/workflows/release_notes_comments_migration.yml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            for (const issueNumber of issueNumbers.split(', ')) {
+              if (issueNumber && commentBody) {
+                await github.rest.issues.createComment({
+                  owner: repoOwner,
+                  repo: repoName,
+                  issue_number: issueNumber,
+                  body: commentBody
+                });            
+              }
+            }
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply .github/workflows/release_notes_comments_migration.yml]: Fixed in 72ed5726450617cb5661ad95cca3138df26f2a8d.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,172 @@
...
+```bash
+git clone https://github.com/AbsaOSS/test-file-suffix-inspector.git
+cd test-file-suffix-inspector
+```
+
+Install the dependencies:
+```bash
+pip install -r requirements.txt
+```
+
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply README.md]: Addressed in 72ed5726450617cb5661ad95cca3138df26f2a8d.

File: README.md (hunk 1 above)
Comment (miroslavpojer) [reply README.md]: Addressed in 72ed5726450617cb5661ad95cca3138df26f2a8d.

File: requirements.txt (hunk 3 above)
Comment (miroslavpojer) [reply requirements.txt]: Addressed in 72ed5726450617cb5661ad95cca3138df26f2a8d.

File: README.md (hunk 2 above)
Comment (Zejnilovic) [reply README.md:108]: Ok, I think I get it, but I don't believe an include should be optional.

Also, do we support some patterns like `*`? Both in include or suffix definition.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,178 @@
...
+## Inputs
+### `suffixes`
+- **Description**: List of suffixes that test files should have, separated by commas.
+- **Required**: Yes
+- **Default**: `UnitTests,IntegrationTests`
+
+### `include_directories`
+- **Description**: List of directories to include in the pattern check. This input limits scanning to these directories only.
+- **Required**: No
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply README.md:31]: There is a default path, "src/test." There is the idea that this setting could be ok for many target projects.

File: README.md (hunk 2 above)
Comment (miroslavpojer) [reply README.md:108]: No * patterns are there in this version.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,178 @@
...
+- **Description**: List of directories to exclude from scanning, separated by commas.
+- **Required**: No
+- **Default**: `dist,node_modules,coverage,target,.idea,.github,.git`
+
+### `exclude_files`
+- **Description**: List of filenames to exclude from suffix checks, separated by commas.
+- **Required**: No
+- **Default**: None
+
+### `case_sensitivity`
```
Comment (Zejnilovic) [inline README.md:44]:
```suggestion
### `case_sensitive`
```

File: README.md [hunk 8]
```diff
@@ -1,2 +1,178 @@
...
+- **Description**: Determines if the filename check should be case-sensitive.
+- **Required**: No
+- **Default**: `true`
+
+### `logic`
```
Comment (Zejnilovic) [inline README.md:49]: `Logic off` or `Logic false` sounds misleading. I am still thinking of replacement word

File: README.md [hunk 9]
```diff
@@ -1,2 +1,178 @@
...
+- **Description**: Specifies the format of the output report. Options include console, csv, and json.
+- **Required**: No
+- **Default**: `console`
+
+### `verbose_logging`
+- **Description**: Enable verbose logging to provide detailed output during the action’s execution, aiding in troubleshooting and setup.
+- **Required**: No
+- **Default**: `false`
+
+### `fail_on_violations`
```
Comment (Zejnilovic) [inline README.md:64]:
```suggestion
### `fail_on_violation`
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,178 @@
...
+- **Description**: Set to true to fail the action if any convention violations are detected. Set to false to continue without failure.
+- **Required**: No
+- **Default**: `false`
+
+## Outputs
+### `conventions_violations`
```
Comment (Zejnilovic) [inline README.md:70]:
```suggestion
### `violation_count`
```

File: README.md [hunk 11]
```diff
@@ -1,2 +1,178 @@
...
+## How It Works
+This action scans the specified `include_directories` for test files and checks if they end with the defined `suffixes` or contain them anywhere in the filename based on the `pattern_logic.` It reports the count of files not meeting the naming conventions, with options to fail the action if violations are found.
+
+## Inputs
+### `suffixes`
+- **Description**: List of suffixes that test files should have, separated by commas.
+- **Required**: Yes
+- **Default**: `UnitTests,IntegrationTests`
+
+### `include_directories`
```
Comment (Zejnilovic) [inline README.md:29]:
```suggestion
### `path`
```

or 

```suggestion
### `paths`
```

File: README.md (hunk 8 above)
Comment (miroslavpojer) [reply README.md:49]: `suffix_logic` == True as default?

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply README.md:31]: If it will be required I can provide example [`src/test/java`, `str/test/scala`].

File: README.md (hunk 7 above)
Comment (miroslavpojer) [reply README.md]: Addressed in commit a52c31f6bfaea253b258aa435f51d01787a1ef9c.

File: README.md (hunk 8 above)
Comment (miroslavpojer) [reply README.md]: Addressed in commit a52c31f6bfaea253b258aa435f51d01787a1ef9c.

File: README.md (hunk 9 above)
Comment (miroslavpojer) [reply README.md]: Addressed in commit a52c31f6bfaea253b258aa435f51d01787a1ef9c.

File: README.md (hunk 10 above)
Comment (miroslavpojer) [reply README.md]: Addressed in commit a52c31f6bfaea253b258aa435f51d01787a1ef9c.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply README.md]: Input renamed and introduced glob logic in commit f4f963226692fbc201573d69a6d9c7332b5d0f9d.

File: README.md (hunk 11 above)
Comment (miroslavpojer) [reply README.md]: Input renamed and introduced glob logic in commit f4f963226692fbc201573d69a6d9c7332b5d0f9d.

File: src/file_suffix_inspector.py [hunk 12]
```diff
@@ -64,17 +61,13 @@ def scan_directory(directory: Path, level: int = 1):
...
-                        scan_directory(item, level + 1)
+                    scan_directory(item, level + 1)
                 else:
                     for include_dir in include_directories:
                         # Check if item path contains the include directory
                         if str(item).__contains__(include_dir):
-                            # CHeck if item is defined as an exception to exclude
-                            if item.name in exclude_files:
+                            # Check if item is defined as an exception to exclude
+                            if item.name in excludes:
```
Comment (Zejnilovic) [inline src/file_suffix_inspector.py]: would this understand global patterns?

File: src/filename_inspector.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    return value
+
+
+def set_output(name: str, value: str):
+    output_file = os.getenv('GITHUB_OUTPUT', None)
+    if output_file:
+        with open(output_file, 'a') as f:
+            f.write(f'{name}={value}\n')
+    else:
+        raise EnvironmentError('GITHUB_OUTPUT environment variable is not set.')
```
Comment (Zejnilovic) [inline src/filename_inspector.py:19]: this looks like a precondition we should check for at the beginning of the script

File: src/filename_inspector.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def set_failed(message: str):
+    print(f'::error::{message}')
+    exit(1)
+
+
+def info(message: str):
+    print(message)
```
Comment (Zejnilovic) [inline src/filename_inspector.py:28]: this adds nothing. Is there a reason to keep it?

File: src/filename_inspector.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    try:
+        suffixes_raw = get_input('suffixes')
+        suffixes = suffixes_raw.split(',') if suffixes_raw else []
+        paths_raw = get_input('paths')
+        paths = paths_raw.split(',')
+        excludes_raw = get_input('excludes')
+        excludes = excludes_raw.split(',')
+        report_format = get_input('report_format')
+        verbose_logging = get_input('verbose_logging') == 'true'
+        fail_on_violation = get_input('fail_on_violation') == 'true'
```
Comment (Zejnilovic) [inline src/filename_inspector.py:69]:
```suggestion
        verbose_logging = get_input('verbose_logging').lower() == 'true'
        fail_on_violation = get_input('fail_on_violation').lower() == 'true'
```

If they write `True` it should be still fine.

File: src/filename_inspector.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
+import fnmatch
+import glob
+import os
+import json
+import csv
+
```
Comment (Zejnilovic) [inline src/filename_inspector.py:6]:
```suggestion
import logging

logger = logging.getLogger(__name__)
```

File: src/filename_inspector.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        report_format = get_input('report_format')
+        verbose_logging = get_input('verbose_logging') == 'true'
+        fail_on_violation = get_input('fail_on_violation') == 'true'
+
+        if verbose_logging:
+            info(f'Suffixes: {suffixes}')
+            info(f'Paths: {paths}')
+            info(f'Excludes: {excludes}')
+            info(f'Report format: {report_format}')
+            info(f'Fail on violations: {fail_on_violation}')
```
Comment (Zejnilovic) [inline src/filename_inspector.py:76]:
```suggestion
            logger.info(f'Suffixes: {suffixes}')
            logger.info(f'Paths: {paths}')
            logger.info(f'Excludes: {excludes}')
            logger.info(f'Report format: {report_format}')
            info(f'Fail on violations: {fail_on_violation}')
```

File: src/filename_inspector.py [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+
+        violations = find_non_matching_files(paths, suffixes, excludes)
+        violation_count = len(violations)
+
+        set_output('violation_count', str(violation_count))
+
+        if report_format == 'console' or verbose_logging:
+            print(f'Total violations: {violation_count}')
+            print(f'Violating files: {violations}')
```
Comment (Zejnilovic) [inline src/filename_inspector.py:85]:
```suggestion
            logger.warning(f'Total violations: {violation_count}')
            logger.warning(f'Violating files: {violations}')
```

File: src/filename_inspector.py [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    output_file = os.getenv('GITHUB_OUTPUT', None)
+    if output_file:
+        with open(output_file, 'a') as f:
+            f.write(f'{name}={value}\n')
+    else:
+        raise EnvironmentError('GITHUB_OUTPUT environment variable is not set.')
+
+
+def set_failed(message: str):
+    print(f'::error::{message}')
```
Comment (Zejnilovic) [inline src/filename_inspector.py:23]:
```suggestion
    logger.warning(message)
```

I've intentionally opted for `warning`; we could even settle on `info` as this is the expected outcome, not an error in the programming.

File: src/filename_inspector.py [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+    exit(1)
+
+
+def info(message: str):
+    print(message)
+
+
+def find_non_matching_files(glob_patterns, suffix_patterns, exclude_patterns):
+    def matches_suffix(file, suffix_patterns):
```
Comment (Zejnilovic) [inline src/filename_inspector.py:32]: Both have defined a variable `suffix_patterns` that is not a good practice. Either use the default parents one and have on less argument in `matches_suffix` or rename it.

File: src/filename_inspector.py [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+        for pattern in suffix_patterns:
+            if fnmatch.fnmatch(file, pattern):
+                return True
+        return False
+
+    def matches_exclude(file, exclude_patterns):
```
Comment (Zejnilovic) [inline src/filename_inspector.py:38]: Same as the above

File: src/filename_inspector.py (hunk 13 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Reworked and tested in commits:
- 16039bf97629247624e790995c7fe698d2cf7094
- 857e388dd30621c4f4bb84aa4c3371453ebf3f61

File: src/filename_inspector.py (hunk 16 above)
Comment (miroslavpojer) [reply src/filename_inspector.py:7]: Used in commit d08c39dc9dd94c5bf27480b7597a8534e8b5ae1b.

File: src/filename_inspector.py (hunk 15 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Applied in commit bd737a91af6446764f3ee9eadc0edd11903f1354.

File: src/filename_inspector.py (hunk 18 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Applied in commit bd737a91af6446764f3ee9eadc0edd11903f1354.

File: src/filename_inspector.py (hunk 19 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Applied in commit bd737a91af6446764f3ee9eadc0edd11903f1354.

Comment (miroslavpojer) [reply src/filename_inspector.py:24]: Returned in commit 994f31093fbfb9e47abced9f7b8d1ad2d4235dfc as this is a common concept to report failure.

File: src/filename_inspector.py (hunk 20 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Fixed in commit 994f31093fbfb9e47abced9f7b8d1ad2d4235dfc.

File: src/filename_inspector.py (hunk 21 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Fixed in commit 994f31093fbfb9e47abced9f7b8d1ad2d4235dfc.

File: src/filename_inspector.py (hunk 19 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Final change in commit 57c06c95c7dad8ac405418f94525273fa0bfb417.

File: README.md [hunk 22]
```diff
@@ -15,26 +15,26 @@
...
 - **Required**: Yes
-- **Default**: `*UnitTest.*,*IntegrationTest.*`
+- **Example**: `*UnitTest.*,*IntegrationTest.*`
 
 ### `paths`
-- **Description**: List of directories to include in the pattern check.
+- **Description**: List of directories to include in the glob pattern check, separated by commas.
 - **Required**: Yes
-- **Default**: `**/src/test/**/*.java,**/src/test/**/*.py`
+- **Example**: `**/src/test/**/*.java,**/src/test/**/*.py`
```
Comment (Zejnilovic) [inline README.md:32]:
```suggestion
- **Example**: `**/src/test/**/*.java,**/src/test/**`
```

Path should be just path to directories. Not with filenames as well.

File: src/filename_inspector.py [hunk 23]
```diff
@@ -1,88 +1,89 @@
...
 
 
-def set_output(name: str, value: str):
-    output_file = os.getenv('GITHUB_OUTPUT', None)
-    if output_file:
-        with open(output_file, 'a') as f:
-            f.write(f'{name}={value}\n')
-    else:
-        raise EnvironmentError('GITHUB_OUTPUT environment variable is not set.')
+def set_output(name: str, value: str, default_output_path: str = "default_output.txt"):
```
Comment (Zejnilovic) [inline src/filename_inspector.py:17]: Where did the raise exception move?

Comment (miroslavpojer) [reply src/filename_inspector.py:17]: A new default value will be returned when the variable is not found in the env vars.

File: src/filename_inspector.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+
+    # Iterate over each glob pattern to find files
+    for path in paths:
+        for file in glob.glob(path, recursive=True):
+            # Check if the file matches any of the exclude patterns
+            if _matches_exclude(file, excludes):
+                continue
+            # Check if the file matches any of the name patterns
+            if not _matches_name_patterns(file, name_patterns):
+                non_matching_files.append(file)
```
Comment (Zejnilovic) [inline src/filename_inspector.py:53]:
```suggestion
            if any(re.match(p, file) for r in excludes):
                continue
            # Check if the file matches any of the name patterns
            if not any(re.match(p, file) for r in name_patterns):
                non_matching_files.append(file)
```

And you could remove 2 functions.

File: README.md (hunk 22 above)
Comment (miroslavpojer) [reply README.md]: Fixed in commit 9b11bb975a4a693842b0858e0f373be0acef0833.

File: src/filename_inspector.py (hunk 24 above)
Comment (miroslavpojer) [reply src/filename_inspector.py]: Used in commit f2f7ad30a9323ebadb92d0ce3cd590837bc973d5.
Instead of re used fnmatch.

**Review by Zejnilovic — APPROVED** [review]

**Pull request merged**